### DISCUSSION TITLE
Server actions, content type?

### DISCUSSION THREAD

**Discussion by anton-johansson** [vercel/next.js#72961] [Help] [answered]
### Summary

We've been exploring [Server Actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations) for a bit. When we deploy our application behind an NGINX with ModSecurity enabled, we realized that it's blocked by this rule:
https://github.com/coreruleset/coreruleset/blob/main/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf#L993

It seems like the Server Actions are called by `POST`'ing data to the server. The data seems to be in JSON format, an array of parameters. But the `Content-Type` header is `
text/plain;charset=UTF-8`. Is this intended? Is there any reason why it couldn't be `application/json`?

Thanks in advance!

**Comment by icyJoseph** [comment, marked as answer]
It depends... I think if you submit a form then it uses, multipart/form-data,, for example:

```js
import { foo } from './actions'

export default function Home() {
  return (
  <form action={foo}>
      <button type="submit">go</button>
    </form>
  );
}
```

Has these headers:

<img width="614" alt="Screenshot 2024-11-19 at 19 35 35" src="https://github.com/user-attachments/assets/feac67cd-64de-4da0-ad9f-4e867cac6b69">

But this:

```js
'use client'
import {foo} from './actions'

export default function Home() {
  return (
      <button  onClick={() => foo()}>go</button>
  );
}
```

Has the headers you report:

<img width="614" alt="Screenshot 2024-11-19 at 19 39 44" src="https://github.com/user-attachments/assets/36808b44-ce25-408d-b0bc-2193fd0aa847">

I think those are the cases. We'd need to loop in React here, since Server Actions are a React feature...

**Reply by icyJoseph** [reply]
And I guess they use that content type because server actions support more types than what the JSON format does.

> As client code calls the Server Action over the network, any arguments passed will need to be serializable.

- https://react.dev/reference/rsc/use-server#serializable-parameters-and-return-values

**Reply by anton-johansson** [reply]
Sorry for the late reply. Thank you both for your answers! I guess we'll simply live with turning off that ModSecurity config. :D

**Reply by revoltez** [reply]
why not allow modifying the Content-Type instead?
